File: Baasket.LogisticsV2/AutomatedTrackingObserver.cs
namespace Baasket.LogisticsV2;

/// <summary>
/// Simulates parcel tracking by advancing the session through timed emoji states:
/// 💳 Paid → 📦 Packed → 🚚 Shipped → ✅ Delivered.
/// </summary>
public sealed class AutomatedTrackingObserver : IObserver, IDisposable
{
    private static readonly CheckoutSessionState[] TrackingSequence =
    [
        CheckoutSessionState.Packed,
        CheckoutSessionState.Shipped,
        CheckoutSessionState.Delivered,
    ];

    private readonly TimeSpan _stepDelay;
    private readonly object _gate = new();
    private readonly Dictionary<Guid, CancellationTokenSource> _activeSimulations = [];

    public AutomatedTrackingObserver(TimeSpan? stepDelay = null)
    {
        _stepDelay = stepDelay ?? TimeSpan.FromSeconds(10);
    }

    public void Update(CheckoutSession session, CheckoutSessionState previousState)
    {
        if (previousState != CheckoutSessionState.CheckoutStarted
            || session.State != CheckoutSessionState.PaymentConfirmed)
        {
            return;
        }

        lock (_gate)
        {
            if (_activeSimulations.ContainsKey(session.SessionId))
                return;

            var cts = new CancellationTokenSource();
            _activeSimulations[session.SessionId] = cts;
            _ = RunSimulationAsync(session, cts.Token);
        }
    }

    private async Task RunSimulationAsync(CheckoutSession session, CancellationToken cancellationToken)
    {
        try
        {
            foreach (var nextState in TrackingSequence)
            {
                await Task.Delay(_stepDelay, cancellationToken).ConfigureAwait(false);
                session.AdvanceTo(nextState);
            }
        }
        catch (OperationCanceledException)
        {
            // Session simulation cancelled — no action required.
        }
        finally
        {
            lock (_gate)
                _activeSimulations.Remove(session.SessionId);
        }
    }

    public void CancelSession(Guid sessionId)
    {
        lock (_gate)
        {
            if (_activeSimulations.TryGetValue(sessionId, out var cts))
            {
                cts.Cancel();
                _activeSimulations.Remove(sessionId);
            }
        }
    }

    public void Dispose()
    {
        lock (_gate)
        {
            foreach (var cts in _activeSimulations.Values)
                cts.Cancel();

            _activeSimulations.Clear();
        }
    }
}

File: logistics_v2/automated_tracking_observer.py
from __future__ import annotations

import threading
import time
from uuid import UUID

from logistics_v2.checkout_session import CheckoutSession
from logistics_v2.checkout_session_state import CheckoutSessionState


class AutomatedTrackingObserver:
    """Timed 💳 Paid → 📦 Packed → 🚚 Shipped → ✅ Delivered simulation."""

    _TRACKING_SEQUENCE = (
        CheckoutSessionState.PACKED,
        CheckoutSessionState.SHIPPED,
        CheckoutSessionState.DELIVERED,
    )

    def __init__(self, step_delay_seconds: float = 10.0) -> None:
        self.step_delay_seconds = step_delay_seconds
        self._gate = threading.Lock()
        self._active: set[UUID] = set()

    def update(
        self,
        session: CheckoutSession,
        previous_state: CheckoutSessionState,
    ) -> None:
        if previous_state is not CheckoutSessionState.CHECKOUT_STARTED:
            return
        if session.state is not CheckoutSessionState.PAYMENT_CONFIRMED:
            return

        with self._gate:
            if session.session_id in self._active:
                return
            self._active.add(session.session_id)

        thread = threading.Thread(
            target=self._run_simulation,
            args=(session.session_id, session),
            daemon=True,
        )
        thread.start()

    def _run_simulation(self, session_id: UUID, session: CheckoutSession) -> None:
        try:
            for next_state in self._TRACKING_SEQUENCE:
                time.sleep(self.step_delay_seconds)
                session.advance_to(next_state)
        except Exception:
            return
        finally:
            with self._gate:
                self._active.discard(session_id)

    def cancel_session(self, session_id: UUID) -> None:
        with self._gate:
            self._active.discard(session_id)
